storagenode/retain/retain.go

```diff
@@ -88,105 +89,207 @@ type Service struct {
 	log    *zap.Logger
 	config Config
 
-	mu     sync.Mutex
-	queued map[storj.NodeID]Request
+	cond    sync.Cond
+	queued  map[storj.NodeID]Request
+	working map[storj.NodeID]struct{}
+	group   errgroup.Group
 
-	reqChan      chan Request
-	semaphore    chan struct{}
-	emptyTrigger chan struct{}
+	closedOnce sync.Once
+	closed     chan struct{}
+	started    bool
 
 	store *pieces.Store
 }
 
 // NewService creates a new retain service.
 func NewService(log *zap.Logger, store *pieces.Store, config Config) *Service {
 	return &Service{
-		log:          log,
-		config:       config,
-		queued:       make(map[storj.NodeID]Request),
-		reqChan:      make(chan Request),
-		semaphore:    make(chan struct{}, config.MaxConcurrentRetain),
-		emptyTrigger: make(chan struct{}),
-		store:        store,
+		log:    log,
+		config: config,
+
+		cond:    *sync.NewCond(&sync.Mutex{}),
+		queued:  make(map[storj.NodeID]Request),
+		working: make(map[storj.NodeID]struct{}),
+		closed:  make(chan struct{}),
+
+		store: store,
 	}
 }
 
 // Queue adds a retain request to the queue.
 // It discards a request for a satellite that already has a queued request.
 // true is returned if the request is queued and false is returned if it is discarded
 func (s *Service) Queue(req Request) bool {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.cond.L.Lock()
+	defer s.cond.L.Unlock()
 
-	// subtract some time to leave room for clock difference between the satellite and storage node
-	req.CreatedBefore = req.CreatedBefore.Add(-s.config.RetainTimeBuffer)
+	select {
+	case <-s.closed:
+		return false
+	default:
+	}
 
-	// only queue retain request if we do not already have one for this satellite
-	if _, ok := s.queued[req.SatelliteID]; !ok {
-		s.queued[req.SatelliteID] = req
-		go func() { s.reqChan <- req }()
+	// subtract some time to leave room for clock difference between the satellite and storage node
+	req.CreatedBefore = req.CreatedBefore.Add(-s.config.MaxTimeSkew)
 
-		return true
-	}
+	s.queued[req.SatelliteID] = req
+	s.cond.Broadcast()
 
-	return false
+	return true
 }
 
 // Run listens for queued retain requests and processes them as they come in.
 func (s *Service) Run(ctx context.Context) error {
-	for {
-		// exit if context has been canceled. Otherwise, block until an item can be added to the semaphore
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case s.semaphore <- struct{}{}:
-		}
+	// Hold the lock while we spawn the workers because a concurrent Close call
+	// can race and wait for them.
+	s.cond.L.Lock()
+	defer s.cond.L.Unlock()
```

Comment:
This will lock for the entire duration of `Run` which I think will cause issues

Reply:
It doesn't. There's an `Unlock` before we `Wait` on the error group.